### PR TITLE
19443 add support for java stream agolubev

### PR DESCRIPTION
**Pull request by agolubev**
- Added stream cancel to Sink.queue.
- Java 8 Stream from Sink.javaStream can close reactive stream when calling stream.close(). 
- Perhaps Sink.javaCollector parameter must be function instead of Collector (to be able to reuse stream)

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 1]
```diff
@@ -31,6 +32,31 @@ object Sink {
     new Sink(scaladsl.Sink.fold[U, In](zero)(f.apply).toCompletionStage())
 
   /**
+   * A `Sink` that will use Java 8 `Collector` to reduce and optionally transform
+   * accumulated result into a final representation.
+   * As collector is mutable you can use it only with one stream.
+   * Reduction operations performs sequentially.
+   */
+  def javaCollector[T, A, R](collector: java.util.stream.Collector[T, A, R]): Sink[T, CompletionStage[R]] =
+    new Sink(scaladsl.Sink.javaCollector[T, A, R](collector).toCompletionStage())
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: did we discuss/decide if these methods should be here or in `StreamConverters`?

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 2]
```diff
@@ -31,6 +32,31 @@ object Sink {
...
+
+  /**
+   * A `Sink` that will use Java 8 `Collector` to reduce and optionally transform
+   * accumulated result into a final representation.
+   * As collector is mutable you can use it only with one stream.
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: and we shouldn't make it factory?

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 1 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: I'll move to SreamConverters

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 2 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Collector was as parameter object in starting point that @drewhk prototyped. Can make it factory function.

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 1 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: @patriknw, after thinking a little. I would left javaCollector in Sink as it's not converting to some java stream it's do transformation inside sink so it's not a converter in pure sense.
However I'll move toJavaStream methods and as JavaStream (for Sink) as they converting from and to Streams.

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Don't you think it will be confusing to not have all in the same place? I need to take another look but AFK at the moment.

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Ok, let me keep them all in `StreamConverters`

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala [hunk 3]
```diff
@@ -133,4 +140,136 @@ class SinkSpec extends AkkaSpec {
...
+
+      // Collector has state so it preserves all previous elements that went though
+      Source(4 to 6).runWith(sink).futureValue should ===(25)
+    }
+
+    "fail if getting the supplier fails" in {
+      def failedSupplier(): Supplier[Array[Int]] = throw TE("")
+      val future = Source(1 to 100).runWith(Sink.javaCollector(
+        new TestCollector(failedSupplier, accumulator, combiner, finisher)))
+      a[TE] shouldBe thrownBy { Await.result(future, 300.millis) }
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala]: This is too short for some of our CI environments. There is no reason to use a short timeout here. Use 3.seconds when in doubt, or better `remaining` or `remainingOrDefault`. Several more occurrences...

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 1 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: um, we discussed this at an Akka Team meeting, but there were a lot of different opinions (naming). I thought we had decided on `StreamConverters`, see this comment: https://github.com/akka/akka/pull/19456#issuecomment-171678612, but it seems like there are still different opinions. We should not rush this one to tomorrow's release, so let's continue discussion next week.

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Ok

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: we should pick up this discussion, @drewhk do you have a different opinion about the placement? You don't want them in `StreamConverters`?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: At that point, where the original PR was opened on which this is based (maybe in January?), there was a discussion, but I don't recall what we ended up with :) I am fine with whatever the majority likes.

**Comment by agolubev** [comment]
Ok, finally pushed all changes according to feedback. Ready for review

**Comment by agolubev** [comment]
Now all java stream related methods are in StreamConverters.
Added factory function as parameter in javaCollector  to achieve reusing of Sink. However this made javaCollector implementation more complex (added function as a wrapper).

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 4]
```diff
@@ -90,4 +96,91 @@ object StreamConverters {
   def asInputStream(readTimeout: FiniteDuration = 5.seconds): Sink[ByteString, InputStream] =
     Sink.fromGraph(new InputStreamSinkStage(readTimeout))
 
+  /**
+    * A `Sink` that will use Java 8 `Collector` to reduce and optionally transform
+    * accumulated result into a final representation.
+    * As collector is mutable you can use it only with one stream.
+    * Reduction operations perform sequentially.
+    */
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: The docs here feels a bit sparse compared to the existing methods, is that because there are no more important things to describe about these methods or could it be improved?

**Comment by johanandren** [comment]
Needs mentions in stages-overview.rst (even if not technically built in stages)

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 5]
```diff
@@ -90,4 +96,91 @@ object StreamConverters {
...
+    * Reduction operations perform in parallel.
+    */
+  def javaCollectorParallelUnordered[T, R](parallelism: Int)(collectorFactory: () ⇒ java.util.stream.Collector[T, _ <: Any, R]): Sink[T, Future[R]] = {
+    if (parallelism == 1) javaCollector[T, R](collectorFactory)
+    else {
+      val createOnceFactory = () => {
+        val ref:AtomicReference[Collector[T, Any, R]] = new AtomicReference[Collector[T, Any, R]]()
+        ref.compareAndSet(null, collectorFactory().asInstanceOf[Collector[T, Any, R]])
+        ref.get()
+      }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: Why is this needed? Couldn't the GraphDSL part just call `collectorFactory()` once and then use that instance to each `CollectorState` it creates?

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala (hunk 4 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: Yeah, will rerwrite scala docs

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala (hunk 5 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: I'm thinking now that GraphDSL part is running in materialization which is exactly what I was trying to achieve. So you are right. I'll fix that

**Comment by agolubev** [comment]
Also several Java tests will be good as well. Will add them

**Comment by agolubev** [comment]
Pushed changes. Ready for review

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 6]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
 asOutputStream
 ^^^^^^^^^^^^^^
 Create a source that materializes into an ``OutputStream``. When bytes are written to the ``OutputStream`` they
-are emitted from the source
+are emitted from the source.
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: thanks! :)

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 7]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
 
 The ``OutputStream`` will no longer be writable when the ``Source`` has been canceled from its downstream, and
 closing the ``OutputStream`` will complete the ``Source``.
 
+asJavaStream
+^^^^^^^^^^^^
+Create a sink which materializes into Java 8 ``Stream`` that can be run to trigger demand through the sink.
+Elements emitted through the stream will be available for reading through the Java ``Stream``.
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: ... into _a_ Java 8  ...

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 8]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
+
+The Java 8 ``Stream`` will be ended when the stream flowing into this ``Sink`` completes, and closing the Java
+``Stream`` will cancel the inflow of this ``Sink``. Java ``Stream`` throws exception in case reactive stream failed.
+
+fromJavaStream
+^^^^^^^^^^^^^^
+Create a source that wraps Java 8 ``Stream``. ``Source`` uses stream iterator to get all its elements and send them
+downstream on demand.
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: ... wraps _a_ Java 8 ...

Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: ... uses _a_ stream iterator ...

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 9]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
+
+javaCollector
+^^^^^^^^^^^^^
+Create a sink which materializes into ``CompletionStage`` which will be completed with a result of the Java 8 Collector
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: ... into _a_ CompletionStage ...

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 10]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
+transformation and reduction operations. This allows to use Java 8 streams transformations for reactive streams.
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: "use" -> "usage of" ?

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 11]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
+Collector is triggering demand downstream. Elements emitted through the stream will be accumulated into a mutable
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: "The Collector will trigger demand downstream.", also, collector is a class so mark it as source code.

File: akka-docs/rst/java/stream/stages-overview.rst [hunk 12]
```diff
@@ -420,11 +420,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
...
+result container, optionally transformed into a final representation after all input elements have been processed.
+Collector can also do reduction at the end. Reduction processing performs sequentially.
```
Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: The

Comment (johanandren) [inline akka-docs/rst/java/stream/stages-overview.rst]: "Reduction processing is performed sequentially"

File: akka-docs/rst/scala/stream/stages-overview.rst [hunk 13]
```diff
@@ -409,11 +409,46 @@ The ``InputStream`` will be closed when the ``Source`` is canceled from its down
 asOutputStream
 ^^^^^^^^^^^^^^
 Create a source that materializes into an ``OutputStream``. When bytes are written to the ``OutputStream`` they
-are emitted from the source
+are emitted from the source.
 
 The ``OutputStream`` will no longer be writable when the ``Source`` has been canceled from its downstream, and
 closing the ``OutputStream`` will complete the ``Source``.
 
+asJavaStream
```
Comment (johanandren) [inline akka-docs/rst/scala/stream/stages-overview.rst]: Same in this section as in the other one!

File: akka-stream-tests/src/test/java/akka/stream/javadsl/SinkTest.java [hunk 14]
```diff
@@ -69,6 +70,22 @@ public void mustBeAbleToUseActorRefSink() throws Exception {
...
+
+  @Test
+  public void mustBeAbleToUseCollector() throws Exception {
+    final List<Integer> list = Arrays.asList(1, 2, 3);
+    final Sink<Integer, CompletionStage<List<Integer>>> collectorSink = StreamConverters.javaCollector(Collectors::toList);
+    CompletionStage<List<Integer>> result = Source.from(list).runWith(collectorSink, materializer);
+    assertEquals(list, result.toCompletableFuture().get(1, TimeUnit.SECONDS));
+  }
+
+  @Test
```
Comment (johanandren) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/SinkTest.java]: 👍

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 15]
```diff
@@ -290,10 +299,16 @@ private[akka] final class SeqStage[T] extends GraphStageWithMaterializedValue[Si
   }
 }
 
+private[stream] object QueueSink {
+  sealed trait Output[+T]
+  final case class Pull[T](promise: Promise[Option[T]]) extends Output[T]
+  case object Cancel extends Output[Nothing]
+}
+
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Hmm, why is this in this PR, isn't this PR about Java Streams?

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: NVM found where it is used waaay below. :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 16]
```diff
@@ -90,4 +96,99 @@ object StreamConverters {
...
+    * elements and send them downstream on demand.
+    *
+    * Example usage: `Source.fromJavaStream(() ⇒ IntStream.rangeClosed(1, 10))`
+    *
+    * You can use [[Source.async]] to create asynchronous boundaries between synchronous Java ``Stream``
+    * and the rest of flow.
+    */
+  def fromJavaStream[T, S <: java.util.stream.BaseStream[T, S]](stream: () ⇒ java.util.stream.BaseStream[T, S]): Source[T, NotUsed] = {
+    import scala.collection.JavaConverters._
+    Source.fromIterator(() ⇒ stream().iterator().asScala)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: `named("fromJavaStream")` ?

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 17]
```diff
@@ -90,4 +96,99 @@ object StreamConverters {
...
+          nextElement = Await.result(nextElementFuture, Inf)
+          nextElement.isDefined
+        }
+
+        override def next(): T = {
+          val next = nextElement.get
+          nextElementFuture = queue.pull()
+          next
+        }
+      }, 0), false).onClose(new Runnable { def run = queue.cancel() })).named("javaStream")
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: "asJavaStream" ?

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 18]
```diff
@@ -90,4 +96,99 @@ object StreamConverters {
...
+    * Java Stream throws exception in case reactive stream failed.
+    */
+  def asJavaStream[T](): Sink[T, java.util.stream.Stream[T]] = {
+    Sink.fromGraph(new QueueSink[T]()).mapMaterializedValue(queue ⇒ StreamSupport.stream(
+      Spliterators.spliteratorUnknownSize(new java.util.Iterator[T] {
+        var nextElementFuture: Future[Option[T]] = queue.pull()
+        var nextElement: Option[T] = null
+
+        override def hasNext: Boolean = {
+          nextElement = Await.result(nextElementFuture, Inf)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: Should this run on a specific dispatcher perhaps?

File: akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala [hunk 19]
```diff
@@ -63,3 +63,14 @@ trait SinkQueue[T] {
    */
   def pull(): CompletionStage[Optional[T]]
 }
+
+/**
+ * This trait adds cancel support to [[SinkQueue]].
+ */
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala]: maybe scaladoc that it is not meant for external implementations?

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 20]
```diff
@@ -366,17 +386,58 @@ final private[stream] class QueueSink[T]() extends GraphStageWithMaterializedVal
...
 }
+
+/**
+ * INTERNAL API
+ *
+ * Helper class to be able to express reduce as a fold for parallel collector
+ */
+private[akka] class ReducerState[T, R](val collector: java.util.stream.Collector[T, Any, R]) {
+  var reduced: Any = null.asInstanceOf[Any]
+  lazy val combiner = collector.combiner()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: both these private?

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 21]
```diff
@@ -366,17 +386,58 @@ final private[stream] class QueueSink[T]() extends GraphStageWithMaterializedVal
...
+}
+
+/**
+ * INTERNAL API
+ *
+ * Helper class to be able to express collection as a fold using mutable data
+ */
+private[akka] class CollectorState[T, R](val collector: java.util.stream.Collector[T, Any, R]) {
+  lazy val accumulated = collector.supplier().get()
+  lazy val accumulator = collector.accumulator()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: both these private?

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 22]
```diff
@@ -366,17 +386,58 @@ final private[stream] class QueueSink[T]() extends GraphStageWithMaterializedVal
...
+  def cancel(): Unit = delegate.cancel()
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * Helper class to be able to express collection as a fold using mutable data
+ */
+private[akka] class CollectorState[T, R](val collector: java.util.stream.Collector[T, Any, R]) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: final?

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 23]
```diff
@@ -366,17 +386,58 @@ final private[stream] class QueueSink[T]() extends GraphStageWithMaterializedVal
...
+
+  def finish(): R = collector.finisher().apply(accumulated)
 }
+
+/**
+ * INTERNAL API
+ *
+ * Helper class to be able to express reduce as a fold for parallel collector
+ */
+private[akka] class ReducerState[T, R](val collector: java.util.stream.Collector[T, Any, R]) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: final?

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala (hunk 18 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: yeah - will rebase this to io dispatcher

**Comment by agolubev** [comment]
@johanandren, thanks for feedback! Will try to make the next review more interesting.

Are there any manual on documentation style or best practices?

**Comment by johanandren** [comment]
Not that I know of, just look at the doco around and try to follow the general style. Language is hard (for me as well as English isn't my first language) so that kind of things are easy to overlook. Thanks for all hard work on this!

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala (hunk 18 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: @agolubev this still need to be fixed, right? This scary aspect should also be mentioned in docs.

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: @patriknw are you thinking about having it non blocking or just another dispatcher?

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: can it even be made non-blocking?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: No, unfortunately not.

File: akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala (hunk 19 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala]: Actually it's available for external usage - as it allow to stop upstream not waiting when it's finished by itself.
Mirroring the SourceQueueWithComplete

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 21 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: only second one - first is used in javaCollectorParallelUnordered

**Comment by agolubev** [comment]
Pushed all changes according to feedback. Let me know if rebase is needed.

File: akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala (hunk 19 above)
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala]: Yeah, I meant its here for outside usage, but not for subtyping, `class MyClass extends SinkQueueWithCancel[Any]`.  Not super important though.

**Comment by johanandren** [comment]
LGTM

**Comment by agolubev** [comment]
Squashed and rebased. Looking for another LGTM.

**Comment by patriknw** [comment]
LGTM
This is done, right?
Do you have any more comments @drewhk or shall we merge?

**Comment by agolubev** [comment]
Rebased just in case.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkAsJavaStreamSpec.scala [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+    "fail if parent stream is failed" in {
+      val javaSource = Source(1 to 100).map(_ ⇒ throw TE("")).runWith(StreamConverters.asJavaStream())
+      a[TE] shouldBe thrownBy {
+        javaSource.findFirst()
+      }
+    }
+
+    "work with collector that is assigned to materialized value" in {
+      val javaSource = Source(1 to 100).map(_.toString).runWith(StreamConverters.asJavaStream())
+      javaSource.collect(Collectors.joining(", ")) should ===((1 to 100).mkString(", "))
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkAsJavaStreamSpec.scala]: nice one

**Comment by drewhk** [comment]
LGTM
